File: Wirtualny_Swiat_Java/Projekt_PO_Java/src/model/World.java
package model;

import model.Organisms.*;

import java.io.*;
import java.util.*;

public class World {
    private int width, height;
    private List<Organism> organisms;
    private List<String> events = new ArrayList<>();
    private final int numberOfOrganisms = 20;
    private final int numberOfTypeOrganisms = 10;
    private final Random random = new Random();
    public boolean endGame = false;
    private int tura = 0;

    public enum Direction {
        GORA, DOL, LEWO, PRAWO
    }
    public Direction humanMove = null;

    public int getTura() {
        return tura;
    }

    public void setTura(int tura) {
        this.tura = tura;
    }

    public World(int width, int height){
        this.width = width;
        this.height = height;
        this.organisms = new ArrayList<>();
    }

    public int getHeight(){
        return this.height;
    }

    public int getWidth(){
        return this.width;
    }

    public boolean isEmpty(Point position){
        for (Organism o : organisms){
            if (o.position.getX() == position.getX() && o.position.getY() == position.getY()){
                return false;
            }
        }
        return true;
    }

    public boolean isCorrectPosition(Point position){
        int x = position.getX();
        int y = position.getY();

        if (x < 0 || y < 0 || x >= width || y>=height){
            return false;
        }

        return true;
    }

    public void addOrganism(Organism org, Point position){
        org.position.setX(position.getX());
        org.position.setY(position.getY());
        organisms.add(org);
    }

    private Organism createRandomOrganism(Point position){
        int randomNum = random.nextInt(numberOfTypeOrganisms);

        switch (randomNum){
            case 0: return new Wolf(this, position);
            case 1: return new Fox(this, position);
            case 2: return new Turtle(this, position);
            case 3: return new Sheep(this, position);
            case 4: return new Antelope(this, position);
//            case 5: return new Cyber_Sheep(this, position);
            case 5: return new BarszczSosnowskiego(this, position);
            case 6: return new Guarana(this, position);
            case 7: return new Grass(this, position);
            case 8: return new Milkweed(this, position);
            case 9: return new Berries(this, position);
        }

        return null;
    }

    public Point getEmptyAndCorrectPosition(){
        Point position;
        do{
            int x = random.nextInt(width);
            int y = random.nextInt(height);
            position = new Point(x, y);
        }while (!isEmpty(position) || !isCorrectPosition(position));

        return position;
    }

    public void createWorld() {
        for (int i = 0; i < numberOfOrganisms; i++){
            Point newPosition = getEmptyAndCorrectPosition();

            Organism newOrganism = createRandomOrganism(newPosition);
            addOrganism(newOrganism, newPosition);
        }

        Point newPosition = getEmptyAndCorrectPosition();
        Organism newOrganism = new Human(this, newPosition);
        addOrganism(newOrganism, newPosition);
    }

    public Human getHuman(){
        for (Organism o : organisms){
            if (o instanceof Human){
                return (Human) o;
            }
        }
        return null;
    }

    public void killHuman(){
        Human human = getHuman();
        if(!human.getImmortality()){
            human.setAlive(false);
            endGame = true;
            addEvent("Człowiek zginął");
        }else{
            addEvent("Człowiek przeżył dzięki aktywowanej nieśmiertelności");
        }
    }

    public int getNumberOfOrganisms() {
        return numberOfOrganisms;
    }

    public List<Organism> getOrganisms() {
        return organisms;
    }

    public void sortOrganisms(){
        organisms.sort(
                Comparator.comparing(Organism::getInitiative)
                        .thenComparing(Organism::getAge).reversed()
        );
    }

    private void cleadDeadOrganisms(){
        organisms.removeIf(o -> !o.getIsAlive());
    }

    public void makeRun(){
        sortOrganisms();
        for (Organism o : new ArrayList<>(organisms)) {
            if (o.getIsAlive()){
                o.Action();
            }
        }
        cleadDeadOrganisms();
    }

    public void moveOrganism(Organism org, Point position){
        org.setPosition(position);
    }

    public void tryToMoveOrganism(Organism org, Point position){
        if (!isCorrectPosition(position)){
            addEvent(org.getNazwa() + " probował wyjść za mape");
        }else{
            if (isEmpty(position)){
                moveOrganism(org, position);
                addEvent(org.getNazwa() + " przesunal sie na (" + position.getX() + ", " + position.getY() + ")");
            }else {
                Organism other = getOrganismaAtPosition(position);
                other.Collision(org);

            }
        }
    }

    public Organism getOrganismaAtPosition(Point position){
        for (Organism o : organisms){
            if (o.getPosition().getX() == position.getX() && o.getPosition().getY() == position.getY()){
                return o;
            }
        }

        return null;
    }

    public void addEvent(String message){
        events.add(message);
    }

    public List<String> getEvents(){
        return events;
    }

    public void clearEventLog(){
        events.clear();
    }

    public boolean isEndGame() {
        return endGame;
    }

    public void saveToFile(){
        try (BufferedWriter writer = new BufferedWriter(new FileWriter("save.txt"))){
            String stringWidth = new String(String.valueOf(width));
            String stringHeight = new String(String.valueOf(height));
            String turaString = new String(String.valueOf(getTura()));
            String immortalityString;
            String immortalityRounds;

            if (getHuman().getImmortality()){
                 immortalityString =  String.valueOf(1);
                 immortalityRounds = String.valueOf(getHuman().getImortalityRounds());
            }else {
                 immortalityString =  String.valueOf(0);
                 immortalityRounds = String.valueOf(0);
            }
            writer.write(stringWidth + " " + stringHeight + " " + turaString);
            writer.newLine();

            Organism human = getHuman();
            writer.write(human.getNazwa() + " " + human.getPosition().getX() + " " + human.getPosition().getY() + " " + human.getPower() + " " + human.getInitiative() + " " + human.getAge() + " " + immortalityString + " " + immortalityRounds);
            writer.newLine();

            for (Organism o : organisms){
                if (o.getIsAlive()){
                    if (o instanceof Human){
                        continue;
                    }
                    writer.write(o.getNazwa() + " " + o.getPosition().getX() + " " + o.getPosition().getY() + " " + o.getPower() + " " + o.getInitiative() + " " + o.getAge());
                    writer.newLine();
                }
            }

        }catch (IOException e){
            e.printStackTrace();
        }
    }

    public void loadGameFromFile(){
        organisms.clear();
        events.clear();
        try (BufferedReader reader = new BufferedReader(new FileReader("save.txt"))){
            String line;

            line = reader.readLine();
            String[] parts = line.split(" ");
            width = Integer.valueOf(parts[0]);
            height = Integer.valueOf(parts[1]);
            tura = Integer.valueOf(parts[2]);

            line = reader.readLine();
            String[] parts3 = line.split(" ");
            int xHuman = Integer.parseInt(parts3[1]);
            int yHuman = Integer.parseInt(parts3[2]);
            int powerHuman = Integer.parseInt(parts3[3]);
            int initiativeHuman = Integer.parseInt(parts3[4]);
            int ageHuman = Integer.parseInt(parts3[5]);
            int immortality = Integer.parseInt(parts3[6]);
            int immortalityRounds = Integer.parseInt(parts3[7]);

            Point pHuman = new Point(xHuman, yHuman);
            Human human = new Human(this, pHuman);
            if (immortality == 1){
                human.setImmortality(true);
                human.setImortalityRounds(immortalityRounds);
            }
            human.setPower(powerHuman);
            human.setInitiative(initiativeHuman);
            human.setAge(ageHuman);
            addOrganism(human, pHuman);


            while ((line = reader.readLine()) != null){

                String[] parts2 = line.split(" ");
                Organism newOrgaznizm;
                int x = Integer.parseInt(parts2[1]);
                int y = Integer.parseInt(parts2[2]);
                int power = Integer.parseInt(parts2[3]);
                int initiative = Integer.parseInt(parts2[4]);
                int age = Integer.parseInt(parts2[5]);
                Point p = new Point(x, y);


                if (Objects.equals(parts2[0], "Wilk")) newOrgaznizm = new Wolf(this, p);
                else if (Objects.equals(parts2[0], "Lis")) newOrgaznizm = new Fox(this, p);
                else if (Objects.equals(parts2[0], "Antylopa")) newOrgaznizm = new Antelope(this, p);
                else if (Objects.equals(parts2[0], "BarszczSosnowskiego")) newOrgaznizm = new BarszczSosnowskiego(this, p);
                else if (Objects.equals(parts2[0], "CyberOwca")) newOrgaznizm = new Cyber_Sheep(this, p);
                else if (Objects.equals(parts2[0], "Guarana")) newOrgaznizm = new Guarana(this, p);
                else if (Objects.equals(parts2[0], "Mlecz")) newOrgaznizm = new Milkweed(this, p);
                else if (Objects.equals(parts2[0], "Owca")) newOrgaznizm = new Sheep(this, p);
                else if (Objects.equals(parts2[0], "Trawa")) newOrgaznizm = new Grass(this, p);
                else if (Objects.equals(parts2[0], "WilczeJagody")) newOrgaznizm = new Berries(this, p);
                else if (Objects.equals(parts2[0], "Zolw")) newOrgaznizm = new Turtle(this, p);
                else {
                    System.out.println("Nieznany organizm przy wczytywaniu pliku" + parts2[0]);
                    continue;
                }

                newOrgaznizm.setPower(power);
                newOrgaznizm.setInitiative(initiative);
                newOrgaznizm.setAge(age);

                addOrganism(newOrgaznizm, p);
            }
        }catch (IOException e){
            e.printStackTrace();;
        }
    }
}

File: WirtualnySwiatPython/world.py
import random
from abc import ABC, abstractmethod

from Animals.Antelope import Antelope
from Animals.CyberSheep import CyberSheep
from Animals.Fox import Fox
from Animals.Human import Human
from Animals.Sheep import Sheep
from Animals.Turtle import Turtle
from Animals.Wolf import Wolf
from Plants.BarszczSosnowskiego import BarszczSosnowskiego
from Plants.Berries import Berries
from Plants.Grass import Grass
from Plants.Guarana import Guarana
from Plants.Milkweed import Milkweed
from point import Point
from organism import Organism
from config import numberOfTypeOrganims, numberOfOrganisms


class World(ABC):
    def __init__(self, width, height):
        self.width = width
        self.height = height
        self.organisms = []
        self.events = []
        self.isEndGame = False
        self.humanMove = None

    def exitGame(self):
        return self.isEndGame

    def setEndGame(self, endGame):
        self.isEndGame = endGame

    def isEmptyPosition(self, position):
        for organism in self.organisms:
            if position == organism.getPosition():
                return False
        return True

    def isCorrectPosition(self, position):
        x, y = position
        if x < 0 or y < 0 or x >= self.width or y >= self.height:
            return False
        return True

    def getEmptyAndCorrectPosition(self):
        while True:
            x = random.randint(0, self.width - 1)
            y = random.randint(0, self.height - 1)
            pos = Point(x, y)
            if (self.isEmptyPosition(pos) and self.isCorrectPosition(pos)): return pos

    def addEvent(self, event):
        self.events.append(event)

    def createRandomOrganism(self, position):
        randomNum = random.randint(0, numberOfTypeOrganims -1)

        if randomNum == 0:
            return Wolf(self, position)
        elif randomNum == 1:
            return Antelope(self, position)
        elif randomNum == 2:
            return BarszczSosnowskiego(self, position)
        elif randomNum == 3:
            return CyberSheep(self, position)
        elif randomNum == 4:
            return Fox(self, position)
        elif randomNum == 5:
            return Sheep(self, position)
        elif randomNum == 6:
            return Turtle(self, position)
        elif randomNum == 7:
            return Berries(self, position)
        elif randomNum == 8:
            return Grass(self, position)
        elif randomNum == 9:
            return Guarana(self, position)
        elif randomNum == 10:
            return Milkweed(self, position)

        return None


    def createWorld(self):
        humanPosition = self.getEmptyAndCorrectPosition()
        self.organisms.append(Human(self, humanPosition))
        for i in range(numberOfOrganisms):
            position = self.getEmptyAndCorrectPosition()
            organism = self.createRandomOrganism(position)
            self.organisms.append(organism)

    def moveOrganism(self, org, position):
        org.setPosition(position)

    def tryToMoveOrganism(self, org, position):
        if not self.isCorrectPosition(position):
            self.addEvent(org.getNazwa() + " probowal wyjsc poza mape")
        else:
            if self.isEmptyPosition(position):
                self.moveOrganism(org, position)
                self.addEvent(f"{org.getNazwa()} przesunal sie na ({position.GetX()}, {position.GetY()})")
            else:
                other = self.getOrganismAtPosition(position)
                other.Collision(org)

    def makeRun(self):
        self.organisms.sort(key=lambda org: (org.getInitiative(), org.getAge()), reverse=True)

        for organism in self.organisms.copy():
            if organism.getIsAlive():
                organism.Action()

        self.cleadDeadOrganism()

    def getHuman(self):
        for organism in self.organisms:
            if isinstance(organism, Human):
                return organism

        return None

    def getOrganismAtPosition(self, position):
        for organism in self.organisms:
            if (organism.getPosition() == position):
                return organism

        return None

    def cleadDeadOrganism(self):
        self.organisms = [organism for organism in self.organisms if organism.getIsAlive()]

    def killHuman(self):
        human = self.getHuman()

        if not human.getImmortality():
            human.setIsAlive(False)
            self.isEndGame = True
            self.addEvent("Czlowiek zginal")
        else:
            self.addEvent("Czlowiek przezyl dzieki aktywowanej niesmiertelnosci")

    def save_to_file(self, filename="save.txt"):
        try:
            with open(filename, "w") as file:
                file.write(f"{self.width} {self.height}\n")

                human = self.getHuman()
                immortality = 1 if human.getImmortality() else 0
                immortality_rounds = human.getImmortalityRounds() if immortality else 0
                file.write(
                    f"{human.getNazwa()} {human.getPosition().GetX()} {human.getPosition().GetY()} {human.getPower()} {human.getInitiative()} {human.getAge()} {immortality} {immortality_rounds}\n")

                for organism in self.organisms:
                    if organism.getIsAlive() and not isinstance(organism, type(human)):
                        file.write(
                            f"{organism.getNazwa()} {organism.getPosition().GetX()} {organism.getPosition().GetY()} {organism.getPower()} {organism.getInitiative()} {organism.getAge()}\n")

            print("Zapisano gre")
        except Exception as e:
            print("Błąd podczas zapisu gry:", e)

    def load_game_from_file(self, filename="save.txt"):
        try:
            self.organisms.clear()
            self.events.clear()

            with open(filename, "r") as file:
                lines = file.readlines()

            w, h = map(int, lines[0].strip().split())
            self.width, self.height = w, h

            # Wczytaj człowieka
            parts = lines[1].strip().split()
            x, y = int(parts[1]), int(parts[2])
            power = int(parts[3])
            initiative = int(parts[4])
            age = int(parts[5])
            immortality = int(parts[6])
            immortality_rounds = int(parts[7])

            from Animals.Human import Human
            p = Point(x, y)
            human = Human(self, p)
            human.setPower(power)
            human.setInitiative(initiative)
            human.setAge(age)
            if immortality:
                human.setImmortality(True)
                human.setImmortalityRounds(immortality_rounds)
            self.organisms.append(human)

            # Wczytaj pozostałe organizmy
            for line in lines[2:]:
                parts = line.strip().split()
                name = parts[0]
                x, y = int(parts[1]), int(parts[2])
                power = int(parts[3])
                initiative = int(parts[4])
                age = int(parts[5])
                p = Point(x, y)

                if name == "Wilk":
                    from Animals.Wolf import Wolf
                    organism = Wolf(self, p)
                elif name == "Lis":
                    from Animals.Fox import Fox
                    organism = Fox(self, p)
                elif name == "Antylopa":
                    from Animals.Antelope import Antelope
                    organism = Antelope(self, p)
                elif name == "BarszczSosnowskiego":
                    from Plants.BarszczSosnowskiego import BarszczSosnowskiego
                    organism = BarszczSosnowskiego(self, p)
                elif name == "CyberOwca":
                    from Animals.CyberSheep import CyberSheep
                    organism = CyberSheep(self, p)
                elif name == "Guarana":
                    from Plants.Guarana import Guarana
                    organism = Guarana(self, p)
                elif name == "Mlecz":
                    from Plants.Milkweed import Milkweed
                    organism = Milkweed(self, p)
                elif name == "Owca":
                    from Animals.Sheep import Sheep
                    organism = Sheep(self, p)
                elif name == "Trawa":
                    from Plants.Grass import Grass
                    organism = Grass(self, p)
                elif name == "WilczeJagody":
                    from Plants.Berries import Berries
                    organism = Berries(self, p)
                elif name == "Zolw":
                    from Animals.Turtle import Turtle
                    organism = Turtle(self, p)
                else:
                    print(f"Nieznany organizm przy wczytywaniu: {name}")
                    continue

                organism.setPower(power)
                organism.setInitiative(initiative)
                organism.setAge(age)
                self.organisms.append(organism)
            print("Wczytano gre")

        except Exception as e:
            print("Błąd podczas wczytywania gry:", e)